Break lines before binary operators

btmills opened the pull request:
<!-- Please provide a brief summary of your changes: -->

~On Friday, @evilebottnawi [requested](https://github.com/prettier/prettier/issues/3806#issuecomment-562603324) a PR to resolve #3806 in v2.0 by breaking lines, where necessary, before binary operators rather than after them, putting the operators at the beginning of each line.~

~Previously, there was special case logic to break lines before pipe operators but after all other operators. I removed the special case handling so that line breaks occur before all operators.~

I left the implementation, tests, and changelog as separate commits to help ease review. Happy to squash or make any other changes as necessary to help get this in!

fixes #3806

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

evilebottnawi commented:
/cc @prettier/core will be great if you look at this

thorn0 commented:
I thought there was a consensus that 2.0 would be focused on deprecating old Node versions etc. Updating to 2.0 should not lead to massive code reformatting. Otherwise too many people might prefer to keep using 1.x.

BrianHVB commented:
The main 2.0 thread(s) have mainly talked about Node versions, but there have also been many comments on not introducing changes that there is disagreement or lack of consensus on. 

But there seems to be a very strong consensus on the proposed fix to #3806. The only argument offered against the change is that it would be breaking and shouldn't be implemented in a minor version. However, for the change itself, there is widespread agreement both in the issue thread and in the JavaScript community. 

AirBnB Issue: https://github.com/airbnb/javascript/issues/1281
AirBnB Style Guidelines: https://github.com/airbnb/javascript#comparison--nested-ternaries

StandardJS (based on ESLint, widely used): https://standardjs.com/rules.html

thorn0 commented:
@BrianHVB You completely miss the point of why the ex-2.0 was renamed to 3.0. I'm not saying that there is no consensus about this change, only that first things first.

thorn0 commented on changelog_unreleased/javascript/pr-7111.md:
```diff
@@ -0,0 +1,43 @@
+
+// Prettier master
+var a
+  = Math.random() * (yRange * (1 - minVerticalFraction))
```
Are we really sure we want to do this with assignments?

btmills replied:
I started with the most straightforward change in hopes of reducing friction, but a few of the comments on #3806 also prefer assignments at the end, so I’m happy to change it if you’d like.

brainkim replied:
If prettier is in favor of operators first formatting (I’m very supportive of this change), one cool thing about putting assignment operators first is that it makes assignments with several binary operators line up in a really satisfying way:

```js
const isEventTarget
  = value != null
  && typeof value.addEventListener === "function"
  && typeof value.removeEventListener === "function"
  && typeof value.dispatchEvent === "function";
```

It might look foreign at first but you get used to it immediately, and it looks a lot nicer than having the first operand on its own line but indented, IMO. It looks even better in the case of chained assignment. Just try it out next time you find yourself writing a long variable declaration with binary operators and you’ll see what I mean.

glen-84 replied:
How can a decision be made here? There is 47/4 voting on @j-f1's [comment](https://github.com/prettier/prettier/issues/3806#issuecomment-360099484), for what that's worth.

You could change it, and see what feedback is received from pre-release versions.

glen-84 replied:
This should probably apply to all [assignment operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Assignment_Operators) (break after, not before).

btmills replied:
It turns out the example here was incorrect, and I corrected it in ba9d939. What I didn't know until I started digging in this evening is that assignment expressions are handled separately, so with this change as-is, assignment operators remain on the previous line, and this change only move binary operators to the beginning of lines!

To verify this, I added explicit tests for the example code here with all of the assignment operators in 57e1bda that show all operators at the end of lines before the logic change and [updated the assignment tests in `32e565d`](https://github.com/prettier/prettier/pull/7111/commits/32e565df9ffe42e50bbdf24411258be9fb342c39#diff-f6af98e5ecd0e5be31b254192fa17b49L90-R141) to show that this change only moves the binary operators and leaves the assignments where they were.

fisker commented:
Anyone considered break line both before and after? 

```js
const isEventTarget = value != null
  && 
  typeof value.addEventListener === "function"
  && 
  typeof value.removeEventListener === "function"
  && 
  typeof value.dispatchEvent === "function"
  && 
  typeof value.dispatchEvent === "function"
  -
  Also_Nice_With_A_Shorter_Operator_Inside
  ?
  SOME_OTHER_EXPRESSION
  ??
  SOME_OTHER_EXPRESSION
```

btmills commented:
I've rebased this change on `next` to resolve merge conflicts, added [tests](https://github.com/prettier/prettier/pull/7111/commits/aadbd29dde4019ff99d0bcb11dfe710d76b216b5#diff-f6af98e5ecd0e5be31b254192fa17b49L93-R156) verifying that this change only moves binary operators to the beginning of lines while leaving assignment operators on their previous lines, and corrected the [changelog](https://github.com/prettier/prettier/blob/300b900e2e279f4dd93474bff10368f603a0e315/changelog_unreleased/javascript/pr-7111.md) in 300b900.

VincentLanglet commented:
@btmills There is still a conflict

phaux commented on tests/logical_expressions/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -13,10 +13,10 @@ const radioSelectedAttr =
+  (isAnyValueSelected
+    && node.getAttribute(radioAttr.toLowerCase()) === radioValue)
+  || (!isAnyValueSelected && values[a].default === true)
+  || a === 0;
```
Would it be possible to also change this case where the multiline operators are nested? When I encounter such a case in my code, I usually refactor it so that prettier doesn't make it ugly like that. If I was writing it manually I would write something like:

```JS
  (
    isAnyValueSelected
    && node.getAttribute(radioAttr.toLowerCase()) === radioValue
  )
  || (
    !isAnyValueSelected 
    && values[a].default === true
  )
  || a === 0;
```

btmills replied:
I'm trying to keep this change as small as possible to improve the odds the team decides to merge it. Based on the volume of :+1:s on #3806, putting binary operators at the beginning of lines has pretty widespread support. It helps that [the logic change](https://github.com/prettier/prettier/pull/7111/files#diff-3abca41914b4b214a2507e33cc54c28bL5725-R5729) to do that exclusively deletes code. I'm concerned that expanding the scope could reduce the chance it gets merged at all.

phaux replied:
Sure. I didn't mean this comment to show as a "review" of this PR. It was just a random question :sweat_smile:

btmills commented:
> @btmills There is still a conflict

@VincentLanglet I'm not surprised one of the snapshots has been updated since I last rebased. Rebasing the branch takes several minutes, so I haven't been actively monitoring this PR to keep it up to date. If the team indicates they're interested in merging, I'll certainly do another rebase as soon as possible!

btmills commented:
I’ve rebased this on top of `master` now that 2.0 has been released.

tattali commented:
Hi, it looks like the conflicting files are just snapshots and this for years.
Are there any plans to merge this pull request into a milestone?

There has been a lot of discussion around this PR I think it is a quite important one.

Thank you!

btmills commented:
I rebased this on `master` after the 2.1.1 release.

Unlike previous rebases, this one required a slight tweak to the logic due to a change in #8491, which was merged since my last rebase. Without the additional tweak, expressions where an object/array/JSX on the right side was inlined would be missing space between the left side and the operator. This case is now handled, and space will be retained before operators that would normally be on a new line but have instead been inlined due to an object/array/JSX on the right side of the binary expression.

```js
// Before tweak to handle shouldInline
const user = renderedUser|| (
    <div>...</div>
);

// Current state of this branch
const user = renderedUser || (
    <div>...</div>
);
```

socsieng commented:
Thanks for putting this PR together. Would love to see it merged.

For some reason, installing the package with the following command does not place `prettier` in the `node_modules/.bin` directory:

```sh
npm install --save-dev prettier@npm:@btmills/prettier@^2.1.1
```

However, installing `prettier` with the following does:

```sh
npm install --save-dev prettier
```

Is this expected behavior?

I worked around this by updating my npm script to reference `node_modules/prettier/bin-prettier.js` directly but it was a little unexpected.

Lavariet commented:
Hi socsieng,

the syntax 
`npm install --save-dev prettier@npm:@btmills/prettier@^2.1.1`
does not work as far as i know since it's for yarn. It would probably be something like (i don't use yarn so i don't know)
`yarn add prettier@npm:@btmills/prettier@^2.1.1`. 

For npm use:
`npm i -D prettier/prettier#pull/7111/head`

socsieng commented:
Thanks @Lavariet, the install from the pull request did the trick (it's significantly slower to install though... thought that it was broken)

```sh
npm i -D prettier/prettier#pull/7111/head
```

FWIW, it installs with the yarn syntax, but doesn't include the executable in the `.bin` folder.

btmills commented:
Rebased after the 2.2.0 release:

- `printBinaryishExpression()` was extracted from `src/language-js/printer-estree.js` to `src/language-js/print/binaryish.js`, but the logic didn't change, so I just had to move the diff.
- The changelog format appears to have changed since the last rebase, so that's now `7111.md` instead of `pr-7111.md`.
- No tests needed to be updated since the last rebase.

---

A demo install via `prettier@npm:@btmills/prettier@^2.2.0` with yarn and npm v7 correctly adds `node_modules/.bin/prettier`. I don't have an npm v6 install handy right now to see if it's been fixed there.

alex12058 commented:
This does not seem to handle comments correctly.

Input:
```js
const logical_expression =
  aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa /* trailing comment */
  /* leading comment */ && bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
  // line comment
  && cccccccccccccccccccccccc
```

Output:
```js
const logical_expression =
  aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa /* trailing comment */
  && /* leading comment */ bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb &&
  // line comment
  cccccccccccccccccccccccc;
```

Leading comments are placed after operators and operators do not appear at the start of the line when there is a line comment.

alex12058 commented:
In `src/language-js/print/binaryish.js` if you replace this:
```js
   right = [
      lineBeforeOperator ? line : "",
      operator,
      lineBeforeOperator ? " " : line,
      rightContent,
      rightSuffix,
    ];
```
with this:
```js
    let comment = hasComment(node.right, CommentCheckFlags.Leading)
      ? rightContent.splice(0, 1)[0]
      : "";
    right = [
      line,
      comment,
      operator,
      " ",
      rightContent,
      rightSuffix,
    ];
```
it will extract the printed comment from `rightContent` and place it before the opearator. This solution is probably too hacky though and I think inline comments should  be placed after the operator to match the `var a = b || /** @type {string} */ (c);` test in `tests/format/js/comments-closure-typecast/__snapshots__/jsfmt.spec.js.snap`.

alex12058 commented:
Using the input from https://github.com/prettier/prettier/pull/7111#issuecomment-1036949402 here are the comment objects generated for the comments. I think there is a bug in `~/work/prettier/src/main/comments.js` that is incorrectly assigning `/* leading comment */` as an `ownLine` comment.

<img width="615" alt="Screen Shot 2022-02-14 at 12 30 28 pm" src="https://user-images.githubusercontent.com/86032026/153950248-666de04a-6f2a-4b5a-ad35-ed8bbbc0af24.png">

Accoding https://github.com/prettier/prettier/blob/main/docs/plugins.md the description for an `ownLine` comment is "if a comment has only whitespace preceding it and a newline afterwards".  I think the `placement` type should be `remaining` as it has code directly following the inline comment (rather than a newline character)

The `isOwnLineComment` function in `src/main/comments.js` returns `hasNewline(text, start, { backwards: true });` This will return `true` if the character at index `start - 1` is a newline character, however it is not checking that the comment ends with a newline. If a comment does not end with a newline character then it is not an `ownLine` comment.

I think return statement for the function `isOwnLineComment` should be changed to:
```js
  return (
    hasNewline(text, start, { backwards: true })
    && hasNewline(text, locEnd(comment))
  );
```

After making this change the `/* leading comment */` comment is correctly assigned the placement value `remaining` but it is incorrectly assigned an an additional trailing comment to `Indentifier aaa...` when it should be assigned as a leading comment to `Identifier bbb...`.
<img width="670" alt="Screen Shot 2022-02-14 at 2 04 49 pm" src="https://user-images.githubusercontent.com/86032026/153958686-2946f163-fb47-402b-b0c6-1c0028af5115.png">

This is because this code enters the `if (precedingNode && followingNode)` block in the code below (from function `attach` in `src/main/comments.js`) as `precedeingNode` is defined and  (its trailing comment) ends with a newline character. The comment seems to assume `precedingNode` is on the same line, which is false if the trailing comment of the `precedingLine` ends with a newline character.
```js
    if (isOwnLineComment(text, options, decoratedComments, index)) {
      ...
    } else if (isEndOfLineComment(text, options, decoratedComments, index)) {
      ...
    } else {
      comment.placement = "remaining";
      if (handleRemainingComment(...args)) {
        // We're good
      } else if (precedingNode && followingNode) {
        // Otherwise, text exists both before and after the comment on
        // the same line. If there is both a preceding and following
        // node, use a tie-breaking algorithm to determine if it should
        // be attached to the next or previous node. In the last case,
        // simply attach the right node;
        const tieCount = tiesToBreak.length;
        if (tieCount > 0) {
          const lastTie = tiesToBreak[tieCount - 1];
          if (lastTie.followingNode !== followingNode) {
            breakTies(tiesToBreak, text, options);
          }
        }
        tiesToBreak.push(context);
      } 
```
<img width="674" alt="Screen Shot 2022-02-14 at 2 48 12 pm" src="https://user-images.githubusercontent.com/86032026/153959743-46a2f940-92c6-4a16-97c7-24b70ad65d45.png">

Summary:
- function `isOwnLineComment` should additionally check that the comment ends with a newline character.
- handling of remaining comments should be fixed.

btmills commented:
Thank you for catching that bug, @alex12058, and thank you even more for suggesting the fix! I was able to use it with a modification to avoid clobbering type annotation comments.

I rebased this on latest `main`.

alex12058 commented:
@btmills Good catch on the type annotation comments.

I found a potential formatting issue with type annotation comments that appear in multline binary operators.

```js
// Input
var a =  veryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongVar || /** @type {string} */
    (c);

// Output
var a =
  veryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongVar
  || /** @type {string} */
  (c);

// Expected
var a =
  veryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongVar
  || /** @type {string} */ (c);
```
[playground link](https://deploy-preview-7111--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEA3AhgJwATuwXm21TkwE8A1Uy6q8um+25p1sgGWgHMKtsAfftgD0AKlHYAAjDIAHONmABnGJgCWULgF9so4QB0oRIgAowASgDchwxhx5CxFoxcM3z9206bemGzEERcSkZeUUVdU0dPWwzK0MQABoQCFkYNWglZFAsTAgAdwAFLAQslHQAG3z0MizkgCNMdDAAazgYAGV0AFs4dg04ZAAzSqU4BqbW9o7ZZo0uZFUAV3GQOG76uAATLe32dE0l9C44ADEITG70GHTNZBB0JZgIJJAACxhuioB1N7V4JSzMBwDqlf5qVD-Mj3MBKOogDRjTAwQpNLhXYajVYAKyUAA8OvMKnAAIpLCDwTEVMbJWaYJH3eroTYVV6ySIwb5qLYwN7IAAcAAZaXkxt8mrJ7uy4EiSK8AI7k+Co1JlB5KAC0UDg222r0wcEVagNqOOGKQI2pqzG3TUi0wK2SSiJpKVgwtWOSMGZXJ5fKQACYvU01BV5gBhCDdc1rJQAVleSzGABVmWVLTS0CsAJJQXawDpgdRpACCeY6MmJVLGWi0QA)

What do you think the output should be?

btmills commented:
@alex12058 I agree with your edit. That seems just far enough away as to be unrelated, so I'm glad this change doesn't affect behavior there either way.

JanJakes commented:
Seeing [Release v3 being planned](https://github.com/prettier/prettier/issues/13142), I'd like to ask whether this is considered for v3. I see it has the 3.0 milestone, but at the same time, it's not on the release checklist. It seems to be the most upvoted PR, issue (second, but almost no downvotes), and comment on the v3 issue. It seems to be non-controversial issue with a broad consensus. Thanks for any information!

fisker commented:
I'm going to postpone this, since we haven't reached a consensus.

sosukesuzuki commented:
This is a major formatting change, so I think we should hide this feature in an experimental flag(Please see https://github.com/prettier/prettier/issues/14527 about experimental flags).

btmills commented:
I'd love to see this merged in, even if it's only experimental initially! Since it sounds like that's likely to happen after v3, I'll wait that's out and rebase this branch again.

fisker commented:
@btmills The team has decided to accept this under `experimentalOperatorPosition` flat, are you willing to implement it?

- #14537 PR try to implement, but @sosukesuzuki seems very busy

btmills commented:
:tada: that's great to hear, and yes, I'm willing to implement it! A couple questions to make sure I'm going in the right direction:

1. Do you prefer `main` or `next` as the base branch?
2. #14537 adds an additional `run_spec(..., { experimentalOperatorPosition: true });` call ([example](https://github.com/prettier/prettier/pull/14537/files#diff-0eeab82bea6ebaa2f14470963e96550380f14e4aa0dee1265191b37a41042fbdR6)) to many of the tests. Should I go with that approach, or would you prefer something more targeted?

fisker commented:
> Do you prefer main or next as the base branch?

`main` branch, since this will be implemented under a flag, we can release it as  a minor version.

> Should I go with that approach, or would you prefer something more targeted?

No, we only need add tests to binary operator related tests. Maybe additional changes from this PR.

btmills commented on src/language-js/options.js:
```diff
@@ -36,6 +36,15 @@ const options = {
     oppositeDescription:
       "Do not print semicolons, except at the beginning of lines which may need them.",
   },
+  experimentalOperatorPosition: {
```
The options defined in this file appeared to follow alphabetical order with a few deviations, so I put this next to and alphabetically before `experimentalTernaries`.

btmills commented on tests/format/js/binary-expressions/__snapshots__/format.test.js.snap:
```diff
@@ -1,5 +1,58 @@
+function f() {
+  const appEntities = getAppEntities(loadObject).filter(
+    (entity) =>
+      entity &&
```
The flag isn't implemented yet as of this commit, so these snapshots just duplicate the default behavior. A later commit that implements the flag's logic will update these tests so the specific changes are easier to see.

btmills commented on src/language-js/print/binaryish.js:
```diff
@@ -267,13 +281,28 @@ function printBinaryishExpressions(
+      }
+      right = [line, comment, operator, " ", rightContent, rightSuffix];
+    } else {
+      right = [
```
This assignment to `right` is just an indentation change from moving it into the `else` branch. Hide whitespace to remove this from the diff.

btmills commented on src/language-js/print/binaryish.js:
```diff
@@ -289,11 +318,15 @@ function printBinaryishExpressions(
       node.left.type !== node.type &&
       node.right.type !== node.type);
+  if (shouldGroup) {
+    right = group(right, { shouldBreak });
```
This line came from the inline expression `shouldGroup ? group(right, { shouldBreak }) : right`. Now the possibly-grouped `right` is used by both branches' `.push()` calls below.

btmills commented on docs/options.md:
```diff
@@ -22,6 +22,17 @@ Valid options:
 | ------- | -------------------------- | ------------------------------- |
 | `false` | `--experimental-ternaries` | `experimentalTernaries: <bool>` |
 
+## Experimental Operator Position
```
Again, I welcome feedback on wording here. I used the Experimental Ternaries section above as a starting point, though I omitted the line about "before it becomes the default behavior" because I don't know whether that's the intent for this flag.

btmills reviewed (commented):
@fisker here you go - rebased on `main` with the logic behind an `experimentalOperatorPosition` flag as requested! I structured the commits to make review easier, so I suggest reviewing one commit at a time. Please let me know if you'd like to see changes!

fisker commented on tests/integration/__tests__/__snapshots__/schema.js.snap:
```diff
@@ -89,6 +89,11 @@ exports[`schema 1`] = `
+          "default": false,
+          "description": "When binary expressions wrap lines, put operators at the start of new lines.",
+          "type": "boolean",
+        },
```
This can't be a boolean, It should be `"line-start" "line-end"` or maybe `"start" | "end"` ?

Or maybe change name to `experimentalOperatorLinebreak`, like [`operator-linebreak`](https://eslint.org/docs/latest/rules/operator-linebreak) rule.

@sosukesuzuki I forgot how we decided.

btmills replied:
@fisker check out 039802c03 – I went with `experimentalOperatorPosition: start|end` since that's the less verbose version. Despite the ESLint rule precedent, I personally find it easier to think about where to place the operator rather than the linebreak relative to the operator. If you all have a strong preference for some other form, I'm happy to tweak it.

fisker replied:
I'm fine with current name.

julien-c replied:
would be awesome to get this merged!

btmills commented:
Through use at work, we discovered that TypeScript union operators were already printed at the beginning of lines by default, but this experimental flag was needed for intersection operators. Following a similar flow as my binary expression commits, 42e05270b adds tests and runs them with `experimentalOperatorPosition: "start"` but does not yet update the logic, then 350c19234 updates the logic and tests so that the impact of the change is easy to review.

storm85049 commented:
Any chance of merging this ? 🥹

btmills commented:
I rebased from 1dfe28f857d5aefa61178cf11aa06ff184d07839 to c7066780ef01b30fe024d64ac10f9d91c6ed0f25 to fix [test failures](https://github.com/prettier/prettier/actions/runs/11993334014?pr=7111) in the most recent CI run related to #16601/#16642. The rebase applied without conflicts, and 32de8ac85b8bc47dfea82b67cf3b702c8e4875f2 was the only change from `yarn test -u`. Since this PR has already been approved, I'll keep 32de8ac85b8bc47dfea82b67cf3b702c8e4875f2 as a separate commit rather than squashing into the prior commits.

sosukesuzuki commented:
CI is failing due to causes unrelated to this PR.
